File: src/main/java/javalin/core/util/RequestUtil.java
// Javalin - https://javalin.io
// Copyright 2017 David Åse
// Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE

package javalin.core.util;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.servlet.http.HttpServletRequest;

import javalin.Request;
import javalin.core.HandlerMatch;

public class RequestUtil {

    public static Request create(HttpServletRequest httpRequest) {
        return new Request(httpRequest, new HashMap<>(), new ArrayList<>());
    }

    public static Request create(HttpServletRequest httpRequest, HandlerMatch handlerMatch) {
        List<String> requestList = Util.pathToList(handlerMatch.requestUri);
        List<String> matchedList = Util.pathToList(handlerMatch.handlerUri);
        return new Request(
            httpRequest,
            getParams(requestList, matchedList),
            getSplat(requestList, matchedList)
        );
    }

    public static List<String> getSplat(List<String> request, List<String> matched) {
        int numRequestParts = request.size();
        int numMatchedParts = matched.size();
        boolean sameLength = (numRequestParts == numMatchedParts);
        List<String> splat = new ArrayList<>();
        for (int i = 0; (i < numRequestParts) && (i < numMatchedParts); i++) {
            String matchedPart = matched.get(i);
            if (isSplat(matchedPart)) {
                StringBuilder splatParam = new StringBuilder(request.get(i));
                if (!sameLength && (i == (numMatchedParts - 1))) {
                    for (int j = i + 1; j < numRequestParts; j++) {
                        splatParam.append("/");
                        splatParam.append(request.get(j));
                    }
                }
                splat.add(urlDecode(splatParam.toString()));
            }
        }
        return Collections.unmodifiableList(splat);
    }

    public static Map<String, String> getParams(List<String> requestPaths, List<String> handlerPaths) {
        Map<String, String> params = new HashMap<>();
        for (int i = 0; (i < requestPaths.size()) && (i < handlerPaths.size()); i++) {
            String matchedPart = handlerPaths.get(i);
            if (isParam(matchedPart)) {
                params.put(matchedPart.toLowerCase(), urlDecode(requestPaths.get(i)));
            }
        }
        return Collections.unmodifiableMap(params);
    }

    public static String urlDecode(String s) {
        try {
            return URLDecoder.decode(s.replace("+", "%2B"), "UTF-8").replace("%2B", "+");
        } catch (UnsupportedEncodingException ignored) {
            return "";
        }
    }

    public static boolean isParam(String pathPart) {
        return pathPart.startsWith(":");
    }

    public static boolean isSplat(String pathPart) {
        return pathPart.equals("*");
    }

    public static String byteArrayToString(byte[] bytes, String encoding) {
        String string;
        if (encoding != null && Charset.isSupported(encoding)) {
            try {
                string = new String(bytes, encoding);
            } catch (UnsupportedEncodingException e) {
                string = new String(bytes);
            }
        } else {
            string = new String(bytes);
        }
        return string;
    }

    public static byte[] toByteArray(InputStream input) throws IOException {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        byte[] byteBuffer = new byte[1024];
        for (int b = input.read(byteBuffer); b != -1; b = input.read(byteBuffer)) {
            baos.write(byteBuffer, 0, b);
        }
        return baos.toByteArray();
    }
}

File: src/main/kotlin/io.javalin/core/util/RequestUtil.kt
/*
 * Javalin - https://javalin.io
 * Copyright 2017 David Åse
 * Licensed under Apache 2.0: https://github.com/tipsy/javalin/blob/master/LICENSE
 *
 */

package io.javalin.core.util

import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.UnsupportedEncodingException
import java.net.URLDecoder
import java.nio.charset.Charset
import java.util.ArrayList
import java.util.Collections
import java.util.HashMap

import javax.servlet.http.HttpServletRequest

import io.javalin.Request
import io.javalin.core.HandlerMatch

object RequestUtil {

    fun create(httpRequest: HttpServletRequest): Request {
        return Request(httpRequest, HashMap<String, String>(), ArrayList<String>())
    }

    fun create(httpRequest: HttpServletRequest, handlerMatch: HandlerMatch): Request {
        val requestList = Util.pathToList(handlerMatch.requestUri)
        val matchedList = Util.pathToList(handlerMatch.handlerUri)
        return Request(
                httpRequest,
                getParams(requestList, matchedList),
                getSplat(requestList, matchedList)
        )
    }

    fun getSplat(request: List<String>, matched: List<String>): List<String> {
        val numRequestParts = request.size
        val numMatchedParts = matched.size
        val sameLength = numRequestParts == numMatchedParts
        val splat = ArrayList<String>()
        var i = 0
        while (i < numRequestParts && i < numMatchedParts) {
            val matchedPart = matched[i]
            if (isSplat(matchedPart)) {
                val splatParam = StringBuilder(request[i])
                if (!sameLength && i == numMatchedParts - 1) {
                    for (j in i + 1..numRequestParts - 1) {
                        splatParam.append("/")
                        splatParam.append(request[j])
                    }
                }
                splat.add(urlDecode(splatParam.toString()))
            }
            i++
        }
        return Collections.unmodifiableList(splat)
    }

    fun getParams(requestPaths: List<String>, handlerPaths: List<String>): Map<String, String> {
        val params = HashMap<String, String>()
        var i = 0
        while (i < requestPaths.size && i < handlerPaths.size) {
            val matchedPart = handlerPaths[i]
            if (isParam(matchedPart)) {
                params.put(matchedPart.toLowerCase(), urlDecode(requestPaths[i]))
            }
            i++
        }
        return Collections.unmodifiableMap(params)
    }

    fun urlDecode(s: String): String {
        try {
            return URLDecoder.decode(s.replace("+", "%2B"), "UTF-8").replace("%2B", "+")
        } catch (ignored: UnsupportedEncodingException) {
            return ""
        }

    }

    fun isParam(pathPart: String): Boolean {
        return pathPart.startsWith(":")
    }

    fun isSplat(pathPart: String): Boolean {
        return pathPart == "*"
    }

    fun byteArrayToString(bytes: ByteArray, encoding: String?): String {
        var string: String
        if (encoding != null && Charset.isSupported(encoding)) {
            try {
                string = String(bytes, Charset.forName(encoding))
            } catch (e: UnsupportedEncodingException) {
                string = String(bytes)
            }

        } else {
            string = String(bytes)
        }
        return string
    }

    @Throws(IOException::class)
    fun toByteArray(input: InputStream): ByteArray {
        val baos = ByteArrayOutputStream()
        val byteBuffer = ByteArray(1024)
        var b = input.read(byteBuffer)
        while (b != -1) {
            baos.write(byteBuffer, 0, b)
            b = input.read(byteBuffer)
        }
        return baos.toByteArray()
    }
}
